core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -30,23 +35,98 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+/**
+ * Class that wraps an Iceberg Catalog to cache table's.
+ *
+ * When `expirationIntervalMillis` is non-zero, the cache will expire tables after
+ * the given interval provided they have not been accessed via the catalog in that time.
+ * Each lookup of the table via the catalog will restart the expiration time.
+ *
+ * If the duration is zero, the cache will retain entries forever, or until
+ * an action is taken on the cache to refresh the cache's table entry.
+ */
 public class CachingCatalog implements Catalog {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CachingCatalog.class);
+  private static final RemovalListener<TableIdentifier, Table> identLoggingRemovalListener =
+      (key, value, cause) -> LOG.debug("{} was evicted from the TableCache with {} cause", key, cause);
+
+  // TODO - Update this to use the default expiration interval millis once things are more complete.
+  //        Don't want to break current tests.
   public static Catalog wrap(Catalog catalog) {
-    return wrap(catalog, true);
+    return wrap(catalog, 0);
+  }
+
+  public static Catalog wrap(Catalog catalog, long expirationIntervalMillis) {
+    return wrap(catalog, true, expirationIntervalMillis);
+  }
+
+  public static Catalog wrap(Catalog catalog, Duration expirationInterval) {
+    return wrap(catalog, true, expirationInterval.toMillis());
   }
 
-  public static Catalog wrap(Catalog catalog, boolean caseSensitive) {
-    return new CachingCatalog(catalog, caseSensitive);
+  public static Catalog wrap(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis) {
+    return new CachingCatalog(catalog, caseSensitive, expirationIntervalMillis);
   }
 
-  private final Cache<TableIdentifier, Table> tableCache = Caffeine.newBuilder().softValues().build();
   private final Catalog catalog;
   private final boolean caseSensitive;
+  private final boolean expirationEnabled;
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected final long expirationIntervalMillis;
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected final Cache<TableIdentifier, Table> tableCache;
+
+  private CachingCatalog(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis) {
+    this(catalog, caseSensitive, expirationIntervalMillis, Ticker.systemTicker());
+  }
 
-  private CachingCatalog(Catalog catalog, boolean caseSensitive) {
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected CachingCatalog(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis, Ticker ticker) {
     this.catalog = catalog;
     this.caseSensitive = caseSensitive;
+    // Set negative values to zero to avoid creating a Duration with a negative value
+    this.expirationIntervalMillis = expirationIntervalMillis <= 0 ? 0 : expirationIntervalMillis;
+    this.expirationEnabled = expirationIntervalMillis > 0;
```

Comment:
Oh I understand. Yeah I will move the logic higher. It would need to be in the specific catalog when we instantiate it (e.g. in SparkCatalog).

I will update.

Reply:
We have to deal with `cache-enabled` because that is existing behavior. But if we use the expiration interval this way, people don't actually have to worry about it in practice. 0 -> don't cache, 1+ -> cache for a while, really big -> don't expire cached tables. That is fairly simple.

Reply:
What happens if:
-  `cache-enabled=true` and `cache.expiration-interval-ms=0`? Should that not cache then?

That's the part that I'm hung up on.

Reply:
As a user, if I had `cache-enabled=true`, I'd be surprised if anything other than caching was happening.

Reply:
Yeah, if you set the cache expiration to immediately then we can skip caching. `cache-enable=true` is the default so it's fine.

Reply:
Ok. I will do that. I will have 1 month be don't expire, 1+ be cache with expire, and 0 be don't cache anymore.

Reply:
I still feel like `-1` would be a better choice for "cache always", with zero turning it off. But I've updated it to 30 days (a relatively large value).

Reply:
We settled on -1 for turning "cache always" (i.e. expiration is off).

Reply:
Oh and the Precondition check ensures that we haven't entered `CachingCatalog` with a value of `0` for expiration interval (as that should turn off caching, which is controlled by no using this wrapper).

Reply:
> We settled on -1 for turning "cache always" (i.e. expiration is off).

Could you explain a bit about the reason the settlement? Or maybe I missed the conversation when reading, if you could link to that. I actually really like the idea of having a max value and treat everything above that as forever. The reasons are:

1. it preserves the meaning of "interval" as non-negative, user does not need to check what -1 means for an interval, like what we are trying to do here.
2. it does not really make sense to keep a cache for a year anyway, a max value does make sense in this case.

And we can just log a warning if the interval value is too large when initializing catalog in engine.